tensorflow/dtensor/mlir/utils/collective_lowering.cc

```diff
@@ -285,6 +285,113 @@ mlir::Operation* EmitTransposeOp(mlir::OpBuilder& builder,
                                                perm_op);
 }
 
+mlir::Operation* EmitCollectiveReduceScatter(
+    mlir::OpBuilder& builder, const mlir::Location& loc, mlir::Value input,
+    mlir::Type output_type, const std::string& reduce_op_str,
+    const mlir::DenseIntElementsAttr& group_assignment,
+    int32 scatter_dimension, int32 key_base,
+    mlir::Value device_id, int32 host_group_size,
+    const mlir::StringRef device_type) {
+  int32_t group_size;
+  mlir::TensorType input_type = input.getType().dyn_cast<mlir::TensorType>();
+
+  const bool need_int32_to_int64_upcast =
+      (device_type.endswith("GPU") && input_type &&
+       input_type.getElementType().isInteger(32));
+
+  const bool need_transpose = scatter_dimension != 0;
+
+  std::vector<int64> perm_for_transpose;
+  if (need_transpose) {
+    perm_for_transpose.reserve(input_type.getRank());
+    for (int i = 0; i < input_type.getRank(); i++) {
+      perm_for_transpose.push_back(i);
+    }
+    std::swap(perm_for_transpose[scatter_dimension], perm_for_transpose[0]);
+    auto pre_transpose_op =
+      EmitTransposeOp(builder, loc, input, perm_for_transpose);
+    input = pre_transpose_op->getResult(0);
+    input_type = input.getType().dyn_cast<mlir::TensorType>();
+    // Compute transposed output type for CollectiveReduceScatter
+    auto output_shape = output_type.dyn_cast<mlir::TensorType>().getShape();
+    std::vector<int64> transposed_shape(output_shape.begin(), output_shape.end());
+    for (int i = 0; i < output_shape.size(); i++) {
+      transposed_shape[i] = output_shape[perm_for_transpose[i]];
+    }
+    output_type = mlir::RankedTensorType::get(
+        transposed_shape, need_int32_to_int64_upcast ? builder.getIntegerType(64) : input_type.getElementType());
+  }
+
+  if (need_int32_to_int64_upcast) {
+    LOG(WARNING) << "On GPU, collective reduce of int32 is not supported. "
+                    "Casting to int64 as a workaround: "
+                 << mlir::debugString(loc);
+
+    mlir::TF::CastOp cast_to_int64 = builder.create<mlir::TF::CastOp>(
+        loc,
+        mlir::RankedTensorType::get(input_type.getShape(),
+                                    builder.getIntegerType(64)),
+        input);
+    input = cast_to_int64.getResult();
+  }
+  mlir::Value group_key_scalar;
```

Comment:
The logic for getting group_key_scalar (and maybe also group_size_scalar are very similar / identical to the other use in the file. Do you see a way of lifting these to a helper function to avoid the redundancy?

Reply:
Thanks @rainwoodman for the review!

Agreed, let me move some of this to a helper function.